### PR TITLE
Platform config improvements

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

The first commit is about making the selection of a "platform" less haphazard. The rest are about trying to clean up exactly what's there so it's a bit less ad-hoc. (I hope to do more on that front later.)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: lib/systems/platforms.nix [hunk 1]
```diff
@@ -469,17 +468,27 @@ rec {
...
-      powerpc64le-linux = powernv;
-    }.${system} or pcBase;
+  # Deprecated 2020-11
+  selectBySystem = system:
+    lib.warn "`selectBySystem` is deprecated use `lib.systems.platforms.select`"
+    select (lib.systems.elaborate { inherit system; });
+
+  select = platform:
+    # x86
+    /**/ if platform.isx86_32 then pc32
```
Comment (matthewbauer) [inline lib/systems/platforms.nix:478]: these should verify we are still looking at linux systems

File: lib/systems/platforms.nix [hunk 2]
```diff
@@ -469,17 +468,27 @@ rec {
...
+    else if platform.isx86_64 then pc64
+
+    # ARM
+    else if platform.isAarch32 then let
+      inherit ({ version = ""; } // platform.parsed.cpu) version;
```
Comment (matthewbauer) [inline lib/systems/platforms.nix:483]:
```suggestion
      version = platform.parsed.cpu.version or "";
```

is more standard

File: lib/systems/platforms.nix [hunk 3]
```diff
@@ -469,17 +468,27 @@ rec {
...
-      armv6l-linux = raspberrypi;
-      armv7a-linux = armv7l-hf-multiplatform;
-      armv7l-linux = armv7l-hf-multiplatform;
-      aarch64-linux = aarch64-multiplatform;
-      mipsel-linux = fuloong2f_n32;
-      powerpc64le-linux = powernv;
-    }.${system} or pcBase;
+  # Deprecated 2020-11
+  selectBySystem = system:
+    lib.warn "`selectBySystem` is deprecated use `lib.systems.platforms.select`"
```
Comment (matthewbauer) [inline lib/systems/platforms.nix:473]: I wonder if anything is using this? Might want to do a quick grep of nixpkgs and other places to make sure we aren't accidentally introducing new warnings. I'd actually prefer if these are internal interfaces that can't be reached outside of nixpkgs (via let bindings or whatever).

File: lib/systems/platforms.nix (hunk 1 above)
Comment (matthewbauer) [reply lib/systems/platforms.nix:478]: Actually the original didn’t check that for pcBase so maybe it shouldn’t matter. I guess defaulting yo pcBase if non linux is reasonable.

File: lib/systems/platforms.nix (hunk 3 above)
Comment (Ericson2314) [reply lib/systems/platforms.nix:473]: That's fine with me. I doubt it's used either.

File: lib/systems/platforms.nix (hunk 1 above)
Comment (Ericson2314) [reply lib/systems/platforms.nix:478]: Yeah I would rather decide what to do with that later. If some non-linux platform has some linux config stuff, it should be harmless.

File: lib/systems/platforms.nix (hunk 3 above)
Comment (Ericson2314) [reply lib/systems/platforms.nix]: It's now removed

**Comment by Ericson2314** [comment]
@matthewbauer OK got all your suggestions.

**Comment by matthewbauer** [comment]
Sorry - forgot to approve - but looks good!

File: pkgs/build-support/bintools-wrapper/default.nix [hunk 4]
```diff
@@ -187,6 +187,7 @@ stdenv.mkDerivation {
       else if targetPlatform.isAlpha then "alpha"
       else if targetPlatform.isVc4 then "vc4"
       else if targetPlatform.isOr1k then "or1k"
+      else if targetPlatform.isRiscV then "lriscv"
```
Comment (JohnAZoidberg) [inline pkgs/build-support/bintools-wrapper/default.nix:190]: Did you test this? I tried to build NixOS on RISC-V a few months ago and I added "riscv" which seemed to work.
I'm now rebasing my branch on top of master and I have a conflict here.

File: lib/systems/platforms.nix [hunk 5]
```diff
@@ -456,10 +456,9 @@ rec {
   ## Other
   ##
 
-  riscv-multiplatform = bits: {
+  riscv-multiplatform = {
     name = "riscv-multiplatform";
     kernelArch = "riscv";
-    bfdEmulation = "elf${bits}lriscv";
```
Comment (Ericson2314) [inline lib/systems/platforms.nix:462]: @JohnAZoidberg it's based off of this which has the "l", presumably for little endian.

**Pull request merged**